components/director/internal/destinationfetchersvc/client.go

```diff
@@ -0,0 +1,212 @@
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("received status code %d when trying to fetch destinations", res.StatusCode)
+	}
+
+	var destinations []model.DestinationInput
+	if err := json.NewDecoder(res.Body).Decode(&destinations); err != nil {
+		return nil, errors.Wrap(err, "failed to decode response body")
+	}
+
+	pageCount := res.Header.Get(c.apiConfig.PagingCountHeader)
+	if pageCount == "" {
+		return nil, errors.Errorf("failed to extract header '%s' from destinations response", c.apiConfig.PagingCountParam)
+	}
+
+	return &DestinationResponse{
+		destinations: destinations,
+		pageCount:    pageCount,
+	}, nil
+}
+
+func (c *Client) buildRequest(ctx context.Context, url string, page string) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to build request")
+	}
+
+	query := req.URL.Query()
+	query.Add(c.apiConfig.PagingCountParam, "true")
+	query.Add(c.apiConfig.PagingPageParam, page)
+	query.Add(c.apiConfig.PagingSizeParam, strconv.Itoa(c.apiConfig.PageSize))
+	req.URL.RawQuery = query.Encode()
+	return req, nil
+}
+
+// FetchDestinationSensitiveData returns sensitive data of a destination
+func (c *Client) FetchDestinationSensitiveData(ctx context.Context, destinationName string) ([]byte, error) {
+	url := fmt.Sprintf("%s%s/%s", c.apiURL, c.apiConfig.EndpointFindDestination, destinationName)
+	log.C(ctx).Infof("Getting destination data from: %s \n", url)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to build request")
+	}
+
+	res, err := c.sendRequestWithRetry(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotFound {
+		return nil, apperrors.NewNotFoundError(resource.Destination, destinationName)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("received status code %d when trying to get destination info for %s",
+			res.StatusCode, destinationName)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read body of response")
+	}
+
+	return body, nil
+}
+
+func (c *Client) sendRequestWithRetry(req *http.Request) (*http.Response, error) {
+	var response *http.Response
+	err := retry.Do(func() error {
+		res, err := c.httpClient.Do(req)
+		if err != nil {
+			return errors.Wrap(err, "failed to execute HTTP request")
+		}
+
+		if err == nil && res.StatusCode < http.StatusInternalServerError {
+			response = res
+			return nil
+		}
+
+		defer res.Body.Close()
+		body, err := ioutil.ReadAll(res.Body)
+
+		if err != nil {
+			return errors.Wrap(err, "failed to read response body")
+		}
+		return errors.Errorf("request failed with status code %d, error message: %v", res.StatusCode, string(body))
+	}, retry.Attempts(c.apiConfig.RetryAttempts), retry.Delay(c.apiConfig.RetryInterval))
+
+	if err != nil {
```

Comment:
return response, err
and remove other code

Reply:
Not that it will be a problem but the response will not be empty in case of error. Which is not a good practice

Reply:
On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed.